### ISSUE TITLE
Cannot use 'import.meta' outside a module

### ISSUE THREAD

**Issue by donalffons** [emscripten-core/emscripten#14311]
Hi,

upgrading from Emscripten 2.0.22 to 2.0.23 breaks the build phase in my project.

Emscripten gives the following error message:
```
/emsdk/upstream/emscripten/tools/acorn-optimizer.js:1845
  throw err;
  ^

SyntaxError: Cannot use 'import.meta' outside a module (1402:51)
  var wasmBinaryFile = new URL('opencascade.wasm', import.meta.url).toString();
                                                   ^

    at Parser.pp$4.raise (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2927:15)
    at Parser.pp$3.parseImportMeta (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2379:14)
    at Parser.pp$3.parseExprImport (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2343:19)
    at Parser.pp$3.parseExprAtom (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2320:21)
    at Parser.pp$3.parseExprSubscripts (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2129:21)
    at Parser.pp$3.parseMaybeUnary (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2106:19)
    at Parser.pp$3.parseExprOps (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2041:21)
    at Parser.pp$3.parseMaybeConditional (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2024:21)
    at Parser.pp$3.parseMaybeAssign (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:1997:21)
    at Parser.pp$3.parseExprList (/emsdk/upstream/emscripten/node_modules/acorn/dist/acorn.js:2830:20) {
  pos: 47015,
  loc: Position { line: 1402, column: 51 },
  raisedAt: 47027
}
emcc: error: '/emsdk/node/14.15.5_64bit/bin/node /emsdk/upstream/emscripten/tools/acorn-optimizer.js /tmp/emscripten_temp_57fzrd80/opencascade.js unsignPointers' failed (returned 1)
```

My build command contains the following arguments:
```js
[
  'emcc',
  '-s',
  'MAIN_MODULE=1',
  '-s',
  'EXPORT_ES6=1',
  '-s',
  'USE_ES6_IMPORT_META=0',
  '-s',
  'USE_FREETYPE=1',
  '-s',
  'EXPORTED_RUNTIME_METHODS=["FS"]',
  '-DIGNORE_NO_ATOMICS=1',
  '-frtti',
  '-fPIC',
  '-DHAVE_RAPIDJSON',
  '-s',
  'DISABLE_EXCEPTION_CATCHING=1',
  '-O3',
  '-s',
  'ALLOW_MEMORY_GROWTH=1',
  '-s',
  'MAXIMUM_MEMORY=4294901760',
  '--bind',
  '/opencascade.js/build/modules/opencascade.js.cpp',
  // many input fils
  '-o',
  '/opencascade.js/dist/opencascade.js'
]
```

**Comment by donalffons** [comment]
...when I remove the flag `USE_ES6_IMPORT_META=0`, the error goes away.

**Comment by RReverser** [comment]
This doesn't make sense to me... that line from the error should be emitted only if `USE_ES6_IMPORT_META=1` (aka the default): https://github.com/emscripten-core/emscripten/blob/3a1fcb84b58f529ec82b272adca4ffd30fdd4cd7/src/preamble.js#L756-L761

Your error looks like #14269 which was _fixed_ in 2.0.23 (in https://github.com/emscripten-core/emscripten/pull/14273), so it seems odd that you only started getting it instead.

Can you print `emcc --version` used to compile that module to be sure?

**Comment by RReverser** [comment]
Actually, upon closer check, it was fixed even earlier, in 2.0.21 -> 2.0.22, so there should be no difference between 2.0.22 and 2.0.23 in terms of behaviour.

**Comment by donalffons** [comment]
Oh well, my bad! That error happened when going from 2.0.21 -> 2.0.22, as you said.
`emcc --version` outputs
```
emcc --version
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 2.0.22 (e998705dbddf40a7f68119e8418b53a5160b19d4)
```

**Comment by RReverser** [comment]
So you're saying you're only getting this error in 2.0.22 but didn't in 2.0.21? That's still odd, because the original reported had it in 2.0.21 and it was fixed in 2.0.22...

**Comment by RReverser** [comment]
Hmm, I can reproduce. Let me recheck few other option combos.

**Comment by donalffons** [comment]
Yes. Here's [my Dockerfile](https://github.com/donalffons/opencascade.js/blob/f71af4da05965b4f63f69492bd6ecba6da57f4d8/Dockerfile#L1), which uses Emscripten version 2.0.22. The corresponding [NPM package](https://www.npmjs.com/package/opencascade.js/v/2.0.0-beta.4fa3125) has the `import.meta` tag inside.

**Comment by RReverser** [comment]
Ah sorry, looks like I was mistaken - info on https://github.com/emscripten-core/emscripten/pull/14273 shows that it made it only into 2.0.23 but didn't make it into the 2.0.22 cut.

Can you please try to upgrade again and recheck?

**Comment by donalffons** [comment]
OK, I'll run another build and let you know when it's done in a couple of hours.

**Comment by RReverser** [comment]
> Can you please try to upgrade again and recheck? (meanwhile double-checking on my end too)

Rechecked, 2.0.23 contains the fix and works as expected.

> when it's done in a couple of hours.

Wow, couple of hours? You must have a really big project (or a slow CI).

**Comment by RReverser** [comment]
That issue aside though, yeah, I'd suggest using that Babel plugin if upgrading Webpack is not an option - I think the ergonomics of the `import.meta` stuff are worth it.

**Comment by donalffons** [comment]
> You must have a really big project (or a slow CI).

It's a big project... CI is not too slow (I'm running this on an N2D GCP compute instance).

I'm having two types of builds. A [full version](https://github.com/donalffons/opencascade.js/runs/2628346991), which bundles everything into one big module. And a [modularized version](https://github.com/donalffons/opencascade.js/runs/2628404147) which creates several MAIN and SIDE_MODULEs. At 3-4 hours, the modularized version is blazingly fast compared to the full version which takes more than 30 hours... The build of the full version annoyingly runs single-threaded for 99% of the time.

Both versions are intended to be used in development, due to the speed and size drawbacks, though... There's a separate build system for creating production builds, which doesn't take so long.

**Comment by RReverser** [comment]
I wonder if there's anything Emscripten could speed up here (that's out of scope of this particular issue, but maybe @kripken would be interested). 30 hours seems like... a lot.

Meanwhile I checked your Dockerfile history, and it makes a bit more sense now: looks like you had 2.0.20 not 2.0.21, which didn't emit `new URL` yet, so going to 2.0.22 would introduce the issue, yeah. One thing I found surprising though is that you seem to have initially updated 2.0.20 -> 2.0.23 (https://github.com/donalffons/opencascade.js/commit/648ab8f40fbc5a80085f47cec0e52b2ae9d144cc) which wouldn't have this issue but then reverted to 2.0.22 instead (https://github.com/donalffons/opencascade.js/commit/a4e89ba327957be60fafc60a35f73d1a406a1d8b) - was there any reason why?

**Comment by donalffons** [comment]
> I'd suggest using that Babel plugin

I have spent quite a bit of time on getting that (and other) plugins to work - but I haven't been successful. The combination of Babel, Webpack, CRA (and CRA-rewired and customize-CRA) is confusing... But I'll try a bit more and will share it here, if I can find a solution.

**Comment by donalffons** [comment]
> One thing I found surprising though is that you seem to have initially updated 2.0.20 -> 2.0.23
eah, there were some bui

**Comment by RReverser** [comment]
I'm sorry, my browser is still struggling through loading of that log 😂

**Comment by RReverser** [comment]
> The problematic part seems to start here:

So that part still bothers me. Both in the tests I've added as part of the fix, and in manual testing I see that 2.0.23 shouldn't even emit that line the error is pointing to. Yet somehow it's in the logs for 2.0.23 upgrade ?

**Comment by RReverser** [comment]
> Yet somehow it's in the logs for 2.0.23 upgrade ?

Ah no, this is also in the same log at the top:

> Status: Downloaded newer image for emscripten/emsdk:2.0.22

So it confirms again that you had an issue with 2.0.22 not 2.0.23.

**Comment by RReverser** [comment]
Anyway, I'm off for now, but so far everything points to 2.0.23 as a fix for your issue. Feel free to close it or comment here once the build is over :)

**Comment by donalffons** [comment]
You're right... The [Dockerfile for that commit](https://github.com/donalffons/opencascade.js/blob/2f1ba69e89d52ce01f24baec9f0d58d2a1c29afe/Dockerfile#L1) states that it's version 2.0.22. Something is weird here - sorry for the confusion.

Anyways, I'm running a local build right now, which should be a bit faster than when going through CI. I'll post here when it's done.

Appreciate your help!

**Comment by sbc100** [comment]
> > You must have a really big project (or a slow CI).
> 
> It's a big project... CI is not too slow (I'm running this on an N2D GCP compute instance).
> 
> I'm having two types of builds. A [full version](https://github.com/donalffons/opencascade.js/runs/2628346991), which bundles everything into one big module. And a [modularized version](https://github.com/donalffons/opencascade.js/runs/2628404147) which creates several MAIN and SIDE_MODULEs. At 3-4 hours, the modularized version is blazingly fast compared to the full version which takes more than 30 hours... The build of the full version annoyingly runs single-threaded for 99% of the time.

I would be curious what is taking all the time in your full build and what you mean by "runs single-threaded".

Are you using LTO?   If not, linking should be fast.

How does your build time compare to the same build time on native toolchains.   Assuming you are not using LTO I would expect roughly comparable times.

If your link options you have:

```
  '/opencascade.js/build/modules/opencascade.js.cpp',
  // many input fils
```

Does this mean you are passing many source files to you link phase?   If you are that would be the first thing to fix.   To get parallism across source files you should be building each source file individually in parallel build system such as `ninja` (or `make -j` if you are old school like that).   I would not even recommend having a single source file in your link step.. that way it easier to measure compile and link time separately and you are maximizing your compile parallelism.  



> 
> Both versions are intended to be used in development, due to the speed and size drawbacks, though... There's a separate build system for creating production builds, which doesn't take so long.

**Comment by donalffons** [comment]
> I would be curious what is taking all the time in your full build and what you mean by "runs single-threaded".

The workflow for building that project goes like this:
1. Pre-processing (15% of build time): My own code uses libclang to analyze the C++ code of the library and auto-generates Embind code and typescript definitions. That Embind code currently has 188473 lines (i.e. it's rather long :wink:)
2. Building with Emscripten (85% of build time): Then Emscripten is taking the Embind files as an input and does its thing.

My own code in phase 1 runs multi-threaded. Phase 2 is using a single thread (i.e. one hyperthreading-core is at 100% load, all others at 0) until the very end. At the end, I can see `wasm-opt` running in parallel (i.e. using all cores), which goes pretty quickly (time depends on the number of cores of the machine).

> Are you using LTO? If not, linking should be fast.

I have been experimenting with LTO builds and found them to be roughly 10% faster. If you want me to do a proper benchmark with both systems, let me know (I can just comment / uncomment some lines [in this Dockerfile](https://github.com/donalffons/opencascade.js/blob/master/Dockerfile#L8)).

> How does your build time compare to the same build time on native toolchains.

Not sure, as I cannot use Embind using a native toolchain.

Currently, I pass "raw" .cpp files to `emcc --bind` during my build. I would expect the builds to be a bit faster, if I compile the `.cpp` files into `.o` and `.a` files first (in parallel) and use those during `emcc --bind`. However, I had some major runtime errors with that approach a while ago, which I wasn't able to resolve. Therefore, I have been using the "raw" `.cpp` approach for the last while.

**Comment by donalffons** [comment]
> I would not even recommend having a single source file in your link step.. that way it easier to measure compile and link time separately and you are maximizing your compile parallelism.

Right, the last paragraph in my comment was about that topic. I will make another build with pre-built `.a` files as input to `emcc --bind` and let you know how much that improves the situation. I'm pretty sure that it won't cut the build times in half, though... As soon as I have those results, I will open a new issue and let you know.

**Comment by sbc100** [comment]
One thing to bear in mind with using `.a` files is that you might want to add `-Wl,--whole-archive` and `-Wl,--no-whole-archive` around the libraries to avoid any semantic differences with compiling directly.  (for example, global constructors in some object could otherwise be ignored with normal use of `.a` files).

**Comment by sbc100** [comment]
> 2\. Building with Emscripten (85% of build time): Then Emscripten is taking the Embind files as an input and does its thing.

I would be very good to get a breakdown of this into compile time and link time.  And the for link time to perhaps break it down further.    But it sounds like you are heading in that direction already.

**Comment by sbc100** [comment]
BTW, passing many source files to `emcc` is bad because (like clang and gcc and all other compilers that I know of) there is no file level parallelism in this case.

**Comment by donalffons** [comment]
> I would be very good to get a breakdown of this into compile time and link time

How can I get a breakdown of this? It doesn't seem to be included in the console output / logs by default. Should I set the environment variable `EMCC_DEBUG=1`?

**Comment by sbc100** [comment]
> > I would be very good to get a breakdown of this into compile time and link time
> 
> How can I get a breakdown of this? It doesn't seem to be included in the console output / logs by default. Should I set the environment variable `EMCC_DEBUG=1`?

Well the compile vs link should be easy to break down (since those are two separate process from your POV).   That is the most important thing to do.. if all goes well you may not need to look further.

If your link time is still very high then we would want to profile the link step.   The best way to do this is to use use `EMPROFILE=1 ./emcc ...`.   You can then run `./emprofile` afterwards and will generate a nice html graph of the breakdown.   Using `EMCC_DEBUG` is not good idea because it can also have performance impacts went enabled.

**Comment by donalffons** [comment]
> everything points to 2.0.23 as a fix for your issue

That did fix it indeed, thanks!

**Comment by donalffons** [comment]
> Well the compile vs link should be easy to break down (since those are two separate process from your POV). That is the most important thing to do..

OK, I see. By "link time" you refer to the time `emcc --bind` takes. Compile time is spent on pre-compiling `.o` and `.a` files. In that case, its easy: Link time is 100% :slightly_smiling_face:.

**Comment by donalffons** [comment]
Thanks a lot for your tipps @sbc100. I'll close this issue here, set up a proper build including your feedback and let you know if I have more problems. I have to go for tonight. Cheers!

**Comment by donalffons** [comment]
@sbc100, quick update regarding the build times:

The full (i.e. non-modularized) build seems to work perfectly fine with the pre-built `.o` files. In a simplified test case for a small subset of the library, the build time is now down to around 1 minute (not including the build time for all `.o` - files, which takes around 1 hours for all source files. That build is cached in a Docker Image, so I'm not concerned about that). The build time with giving `.cpp` files as inputs was at around 90 minutes for that same build configuration.

I'm already super happy with these results. Next, I'll try the pre-compiled `.o` files with the modularized version along with your tip from #13241 (pre-compiling the Embind bindings) and hopefully that'll work too.

Thanks a lot!

**Comment by donalffons** [comment]
@sbc100, just a quick follow up. Big thanks to your hints on how to improve the build system. The build time went from more than 30 hours down to "just" 4:30 hours on a GCP C8 compute instance with 16 threads.

![2021-07-03_13-10](https://user-images.githubusercontent.com/33040347/124352385-360f1180-dc00-11eb-9128-c3d81c79e390.png)

The most amount of time is now spent in the compilation step, where all Embind code (in `.o` files) is compiled to `.o` files. The final linking / build step is so short that it doesn't even show up in the graph any more (it's 3 minutes).

Another quick win will be to parallelize my own code which generates Embind and Typescript code (currently takes ~45 minutes).

All in all, this is a huge improvement. Thanks a lot!
